query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,271 @@
+  }
+
+  private static pruneRefinedTypes(
+    toPrune: Scope | undefined,
+    refiningType: GraphQLCompositeType
+  ) : Scope | undefined {
+    if (!toPrune) {
+      return undefined;
+    }
+    if (!toPrune.directives && isTypeSubTypeOf(toPrune.context.schema, refiningType, toPrune.parentType)) {
+      // The newly added type is a subtype of the current "link", and the current link has no directives,
+      // so it's not useful anymore. Skip it, and check if we can prune further.
+      return Scope.pruneRefinedTypes(toPrune.enclosing, refiningType);
+    }
+    return new Scope(
+      toPrune.context,
+      toPrune.parentType,
+      toPrune.directives,
+      Scope.pruneRefinedTypes(toPrune.enclosing, refiningType)
+    );
+  }
+
+  /**
+   * Whether refining this scope by the provided type would strictly reduce the possible runtime types
+   * of the scope.
+   */
+  isStrictlyRefinedBy(type: GraphQLCompositeType) : boolean {
+    // If one of the type in the chain is a subtype of the provided type, then the type will not strictly
+    // refine the scope.
+    let scope: Scope | undefined = this;
+    while (scope) {
+      if (isTypeSubTypeOf(this.context.schema, scope.parentType, type)) {
+        return false;
+      }
+      scope = scope.enclosing;
+    }
+    return true;
+  }
+
+  /**
+   *  Whether this scope is restricting the possible runtime types of the provided type.
+   */
+  isStrictlyRefining(type: GraphQLCompositeType) : boolean {
+    // This scope will refine the provided type, unless that provided type is a subtype of all
+    // the type in the chain.
+    let scope: Scope | undefined = this;
+    while (scope) {
+      if (scope.parentType !== type && isTypeSubTypeOf(this.context.schema, scope.parentType, type)) {
+        return true;
+      }
+      scope = scope.enclosing;
+    }
+    return false;
+  }
+
+  private computePossibleRuntimeTypes() : ReadonlyArray<GraphQLObjectType> {
+    // The possible runtime types is the intersection of all the possible types of each condition in scope.
+    let possibleTypes = this.context.getPossibleTypes(this.parentType);
+    let nextScope = this.enclosing;
+    while (nextScope) {
+      let enclosingPossibleTypes = this.context.getPossibleTypes(nextScope.parentType);
+      possibleTypes = possibleTypes.filter(t => enclosingPossibleTypes.includes(t));
+      nextScope = nextScope.enclosing;
+    }
+    return possibleTypes;
+  }
+
+  /**
+   * Computes the set of object types that an object within this scope can have at runtime.
+   */
+  possibleRuntimeTypes() : ReadonlyArray<GraphQLObjectType> {
+    if (!this.cachedRuntimeTypes) {
+      this.cachedRuntimeTypes = this.computePossibleRuntimeTypes();
+    }
+    return this.cachedRuntimeTypes;
+  }
+
+  private static valueIdentityKey(value: ValueNode) : string {
+    // We distinguish different value types though a leading character with a comma (so the int 3 has key "i'3") to
```

Comment:
This character is generally called a single quote or apostrophe in English; `,` is a comma.

Reply:
Oops. Fixed.